README.md
# rubyzip
[![Gem Version](https://badge.fury.io/rb/rubyzip.svg)](http://badge.fury.io/rb/rubyzip)
[![Build Status](https://secure.travis-ci.org/rubyzip/rubyzip.svg)](http://travis-ci.org/rubyzip/rubyzip)
[![Code Climate](https://codeclimate.com/github/rubyzip/rubyzip.svg)](https://codeclimate.com/github/rubyzip/rubyzip)
[![Coverage Status](https://img.shields.io/coveralls/rubyzip/rubyzip.svg)](https://coveralls.io/r/rubyzip/rubyzip?branch=master)

Rubyzip is a ruby library for reading and writing zip files.

## Important note

The Rubyzip interface has changed!!! No need to do `require "zip/zip"` and `Zip` prefix in class names removed.

If you have issues with any third-party gems that require an old version of rubyzip, you can use this workaround:

```ruby
gem 'rubyzip', '>= 1.0.0' # will load new rubyzip version
gem 'zip-zip' # will load compatibility for old rubyzip API.
```

## Requirements

* Ruby 1.9.2 or greater

## Installation
Rubyzip is available on RubyGems:

```
gem install rubyzip
```

Or in your Gemfile:

```ruby
gem 'rubyzip'
```

## Usage

### Basic zip archive creation

```ruby
require 'rubygems'
require 'ruby_zip'

folder = "Users/me/Desktop/stuff_to_zip"
input_filenames = ['image.jpg', 'description.txt', 'stats.csv']

zipfile_name = "/Users/me/Desktop/archive.zip"

RubyZip::File.open(zipfile_name, RubyZip::File::CREATE) do |zipfile|
  input_filenames.each do |filename|
    # Two arguments:
    # - The name of the file as it will appear in the archive
    # - The original file, including the path to find it
    zipfile.add(filename, folder + '/' + filename)
  end
  zipfile.get_output_stream("myFile") { |os| os.write "myFile contains just this" }
end
```

### Zipping a directory recursively
Copy from [here](https://github.com/rubyzip/rubyzip/blob/05916bf89181e1955118fd3ea059f18acac28cc8/samples/example_recursive.rb )

```ruby
require 'ruby_zip'

# This is a simple example which uses rubyzip to
# recursively generate a zip file from the contents of
# a specified directory. The directory itself is not
# included in the archive, rather just its contents.
#
# Usage:
#   directory_to_zip = "/tmp/input"
#   output_file = "/tmp/out.zip"
#   zf = ZipFileGenerator.new(directory_to_zip, output_file)
#   zf.write()
class ZipFileGenerator
  # Initialize with the directory to zip and the location of the output archive.
  def initialize(input_dir, output_file)
    @input_dir = input_dir
    @output_file = output_file
  end

  # Zip the input directory.
  def write
    entries = Dir.entries(@input_dir) - %w(. ..)

    ::RubyZip::File.open(@output_file, ::RubyZip::File::CREATE) do |io|
      write_entries entries, '', io
    end
  end

  private

  # A helper method to make the recursion work.
  def write_entries(entries, path, io)
    entries.each do |e|
      zip_file_path = path == '' ? e : File.join(path, e)
      disk_file_path = File.join(@input_dir, zip_file_path)
      puts "Deflating #{disk_file_path}"

      if File.directory? disk_file_path
        recursively_deflate_directory(disk_file_path, io, zip_file_path)
      else
        put_into_archive(disk_file_path, io, zip_file_path)
      end
    end
  end

  def recursively_deflate_directory(disk_file_path, io, zip_file_path)
    io.mkdir zip_file_path
    subdir = Dir.entries(disk_file_path) - %w(. ..)
    write_entries subdir, zip_file_path, io
  end

  def put_into_archive(disk_file_path, io, zip_file_path)
    io.get_output_stream(zip_file_path) do |f|
      f.write(File.open(disk_file_path, 'rb').read)
    end
  end
end
```

### Save zip archive entries in sorted by name state

To save zip archives in sorted order like below, you need to set `::RubyZip.sort_entries` to `true`

```
Vegetable/
Vegetable/bean
Vegetable/carrot
Vegetable/celery
fruit/
fruit/apple
fruit/kiwi
fruit/mango
fruit/orange
```

After this, entries in the zip archive will be saved in ordered state.

### Default permissions of zip archives

On Posix file systems the default file permissions applied to a new archive
are (0666 - umask), which mimics the behavior of standard tools such as `touch`.

On Windows the default file permissions are set to 0644 as suggested by the
[Ruby File documentation](http://ruby-doc.org/core-2.2.2/File.html).

When modifying a zip archive the file permissions of the archive are preserved.

### Reading a Zip file

```ruby
RubyZip::File.open('foo.zip') do |zip_file|
  # Handle entries one by one
  zip_file.each do |entry|
    # Extract to file/directory/symlink
    puts "Extracting #{entry.name}"
    entry.extract(dest_file)

    # Read into memory
    content = entry.get_input_stream.read
  end

  # Find specific entry
  entry = zip_file.glob('*.csv').first
  puts entry.get_input_stream.read
end
```

#### Notice about ::RubyZip::InputStream

`::RubyZip::InputStream` usable for fast reading zip file content because it not read Central directory.

But there is one exception when it is not working - General Purpose Flag Bit 3.

```
If bit 3 (0x08) of the general-purpose flags field is set, then the CRC-32 and file sizes are not known when the header is written. The fields in the local header are filled with zero, and the CRC-32 and size are appended in a 12-byte structure (optionally preceded by a 4-byte signature) immediately after the compressed data
```

If `::RubyZip::InputStream` finds such entry in the zip archive it will raise an exception.

### Password Protection (Experimental)

Rubyzip supports reading/writing zip files with traditional zip encryption (a.k.a. "ZipCrypto"). AES encryption is not yet supported. It can be used with buffer streams, e.g.:

```ruby
RubyZip::OutputStream.write_buffer(::StringIO.new(''), RubyZip::TraditionalEncrypter.new('password')) do |out|
  out.put_next_entry("my_file.txt")
  out.write my_data
end.string
```

This is an experimental feature and the interface for encryption may change in future versions.

## Known issues

### Modify docx file with rubyzip

Use `write_buffer` instead `open`. Thanks to @jondruse

```ruby
buffer = RubyZip::OutputStream.write_buffer do |out|
  @zip_file.entries.each do |e|
    unless [DOCUMENT_FILE_PATH, RELS_FILE_PATH].include?(e.name)
      out.put_next_entry(e.name)
      out.write e.get_input_stream.read
     end
  end

  out.put_next_entry(DOCUMENT_FILE_PATH)
  out.write xml_doc.to_xml(:indent => 0).gsub("\n","")

  out.put_next_entry(RELS_FILE_PATH)
  out.write rels.to_xml(:indent => 0).gsub("\n","")
end

File.open(new_path, "wb") {|f| f.write(buffer.string) }
```

## Configuration

By default, rubyzip will not overwrite files if they already exist inside of the extracted path.  To change this behavior, you may specify a configuration option like so:

```ruby
RubyZip.on_exists_proc = true
```

If you're using rubyzip with rails, consider placing this snippet of code in an initializer file such as `config/initializers/rubyzip.rb`

Additionally, if you want to configure rubyzip to overwrite existing files while creating a .zip file, you can do so with the following:

```ruby
RubyZip.continue_on_exists_proc = true
```

If you want to store non-english names and want to open them on Windows(pre 7) you need to set this option:

```ruby
RubyZip.unicode_names = true
```

Some zip files might have an invalid date format, which will raise a warning. You can hide this warning with the following setting:

```ruby
RubyZip.warn_invalid_date = false
```

You can set the default compression level like so:

```ruby
RubyZip.default_compression = Zlib::DEFAULT_COMPRESSION
```
It defaults to `Zlib::DEFAULT_COMPRESSION`. Possible values are `Zlib::BEST_COMPRESSION`, `Zlib::DEFAULT_COMPRESSION` and `Zlib::NO_COMPRESSION`

You can set multiple settings at the same time by using a block:

```ruby
  RubyZip.setup do |c|
    c.on_exists_proc = true
    c.continue_on_exists_proc = true
    c.unicode_names = true
    c.default_compression = Zlib::BEST_COMPRESSION
  end
```

By default, Zip64 support is disabled for writing. To enable it do this:

```ruby
RubyZip.write_zip64_support = true
```

_NOTE_: If you will enable Zip64 writing then you will need zip extractor with Zip64 support to extract archive.

## Developing

To run the test you need to do this:

```
bundle install
rake
```

## Website and Project Home

http://github.com/rubyzip/rubyzip

http://rdoc.info/github/rubyzip/rubyzip/master/frames

## Authors

Alexander Simonov ( alex at simonov.me)

Alan Harper ( alan at aussiegeek.net)

Thomas Sondergaard (thomas at sondergaard.cc)

Technorama Ltd. (oss-ruby-zip at technorama.net)

extra-field support contributed by Tatsuki Sugiura (sugi at nemui.org)

## License

Rubyzip is distributed under the same license as ruby. See
http://www.ruby-lang.org/en/LICENSE.txt


Changelog.md
1.2.2
=====

* Renamed Zip module to RubyZip. All tests passed in Ruby v2.2.7

1.2.1
=====

* Add accessor to @internal_file_attributes #304
* Extended globbing #303
* README updates #283, #289
* Cleanup after tests #298, #306
* Fix permissions on new zip files #294, #300
* Fix examples #297
* Support cp932 encoding #308
* Fix Directory traversal vulnerability #315
* Allow open_buffer to work without a given block #314

1.2.0
=====

* Don't enable JRuby objectspace #252
* Fixes an exception thrown when decoding some weird .zip files #248
* Use duck typing with IO methods #244
* Added error for empty (zero bit) zip file #242
* Accept StringIO in RubyZip.open_buffer #238
* Do something more expected with new file permissions #237
* Case insensitivity option for #find_entry #222
* Fixes in documentation and examples

1.1.7
=====

* Fix UTF-8 support for comments
* `RubyZip.sort_entries` working for zip output
* Prevent tempfile path from being unlinked by garbage collection
* NTFS Extra Field (0x000a) support
* Use String#tr instead of String#gsub
* Ability to not show warning about incorrect date
* Be smarter about handling buffer file modes.
* Support for Traditional Encryption (ZipCrypto)

1.1.6
=====

* Revert "Return created zip file from RubyZip::File.open when supplied a block"

1.1.5
=====

* Treat empty file as non-exists (@layerssss)
* Revert regression commit
* Return created zip file from RubyZip::File.open when supplied a block (@tpickett66)
* RubyZip::Entry::DEFLATED is forced on every file (@mehmetc)
* Add InputStream#ungetc (@zacstewart)
* Alias for legacy error names (@orien)

1.1.4
=====

* Don't send empty string to stream (@mrloop)
* RubyZip::Entry::DEFLATED was forced on every file (@mehmetc)
* Alias for legacy error names (@orien)

1.1.3
=====

* Fix compatibility of ::OutputStream::write_buffer (@orien)
* Clean up tempfiles from output stream (@iangreenleaf)

1.1.2 
=====

* Fix compatibility of ::RubyZip::File.write_buffer

1.1.1
=====

* Speedup deflater (@loadhigh)
* Less Arrays and Strings allocations (@srawlins)
* Fix Zip64 writting support (@mrjamesriley)
* Fix StringIO support (@simonoff)
* Posibility to change default compression level
* Make Zip64 write support optional via configuration 

1.1.0
=====

* StringIO Support
* Zip64 Support
* Better jRuby Support
* Order of files in the archive can be sorted
* Other small fixes

1.0.0
=====

* Removed support for Ruby 1.8
* Changed the API for gem. Now it can be used without require param in Gemfile.
* Added read-only support for Zip64 files.
* Added support for setting Unicode file names.

0.9.9
=====

* Added support for backslashes in zip files (generated by the default Windows zip packer for example) and comment sections with the comment length set to zero even though there is actually a comment.

0.9.8
=====

* Fixed: "Unitialized constant NullInputStream" error

0.9.5
=====

* Removed support for loading ruby in zip files (ziprequire.rb).

0.9.4
=====

* Changed ZipOutputStream.put_next_entry signature (API CHANGE!). Now allows comment, extra field and compression method to be specified.

0.9.3
=====

* Fixed: Added ZipEntry::name_encoding which retrieves the character
encoding of the name and comment of the entry.
* Added convenience methods ZipEntry::name_in(enc) and ZipEntry::comment_in(enc) for
getting zip entry names and comments in a specified character
encoding.

0.9.2
=====

* Fixed: Renaming an entry failed if the entry's new name was a different length than its old name. (Diego Barros)

0.9.1
=====

* Added symlink support and support for unix file permissions. Reduced memory usage during decompression.
* New methods ZipFile::[follow_symlinks, restore_times, restore_permissions, restore_ownership].
* New methods ZipEntry::unix_perms, ZipInputStream::eof?.
* Added documentation and test for new ZipFile::extract.
* Added some of the API suggestions from sf.net #1281314.
* Applied patch for sf.net bug #1446926.
* Applied patch for sf.net bug #1459902.
* Rework ZipEntry and delegate classes.

0.5.12
======

* Fixed problem with writing binary content to a ZipFile in MS Windows.

0.5.11
======

* Fixed name clash file method copy_stream from fileutils.rb. Fixed problem with references to constant CHUNK_SIZE.
* ZipInputStream/AbstractInputStream read is now buffered like ruby IO's read method, which means that read and gets etc can be mixed. The
  unbuffered read method has been renamed to sysread.

0.5.10
======

* Fixed method name resolution problem with FileUtils::copy_stream and IOExtras::copy_stream.

0.5.9
=====

* Fixed serious memory consumption issue

0.5.8
=====

* Fixed install script.

0.5.7
=====

lib/ruby_zip/entry.rb
module RubyZip
  class Entry
    STORED   = 0
    DEFLATED = 8
    # Language encoding flag (EFS) bit
    EFS = 0b100000000000

    attr_accessor :comment, :compressed_size, :crc, :extra, :compression_method,
                  :name, :size, :local_header_offset, :zipfile, :fstype, :external_file_attributes,
                  :internal_file_attributes,
                  :gp_flags, :header_signature, :follow_symlinks,
                  :restore_times, :restore_permissions, :restore_ownership,
                  :unix_uid, :unix_gid, :unix_perms,
                  :dirty
    attr_reader :ftype, :filepath # :nodoc:

    def set_default_vars_values
      @local_header_offset      = 0
      @local_header_size        = nil # not known until local entry is created or read
      @internal_file_attributes = 1
      @external_file_attributes = 0
      @header_signature         = ::RubyZip::CENTRAL_DIRECTORY_ENTRY_SIGNATURE

      @version_needed_to_extract = VERSION_NEEDED_TO_EXTRACT
      @version                   = VERSION_MADE_BY

      @ftype           = nil          # unspecified or unknown
      @filepath        = nil
      @gp_flags        = 0
      if ::RubyZip.unicode_names
        @gp_flags |= EFS
        @version = 63
      end
      @follow_symlinks = false

      @restore_times       = true
      @restore_permissions = false
      @restore_ownership   = false
      # BUG: need an extra field to support uid/gid's
      @unix_uid            = nil
      @unix_gid            = nil
      @unix_perms          = nil
      # @posix_acl = nil
      # @ntfs_acl = nil
      @dirty               = false
    end

    def check_name(name)
      return unless name.start_with?('/')
      raise ::RubyZip::EntryNameError, "Illegal ZipEntry name '#{name}', name must not start with /"
    end

    def initialize(*args)
      name = args[1] || ''
      check_name(name)

      set_default_vars_values
      @fstype = ::RubyZip::RUNNING_ON_WINDOWS ? ::RubyZip::FSTYPE_FAT : ::RubyZip::FSTYPE_UNIX

      @zipfile            = args[0] || ''
      @name               = name
      @comment            = args[2] || ''
      @extra              = args[3] || ''
      @compressed_size    = args[4] || 0
      @crc                = args[5] || 0
      @compression_method = args[6] || ::RubyZip::Entry::DEFLATED
      @size               = args[7] || 0
      @time               = args[8] || ::RubyZip::DOSTime.now

      @ftype = name_is_directory? ? :directory : :file
      @extra = ::RubyZip::ExtraField.new(@extra.to_s) unless @extra.is_a?(::RubyZip::ExtraField)
    end

    def time
      if @extra['UniversalTime']
        @extra['UniversalTime'].mtime
      elsif @extra['NTFS']
        @extra['NTFS'].mtime
      else
        # Standard time field in central directory has local time
        # under archive creator. Then, we can't get timezone.
        @time
      end
    end

    alias mtime time

    def time=(value)
      unless @extra.member?('UniversalTime') || @extra.member?('NTFS')
        @extra.create('UniversalTime')
      end
      (@extra['UniversalTime'] || @extra['NTFS']).mtime = value
      @time                         = value
    end

    def file_type_is?(type)
      raise InternalError, "current filetype is unknown: #{inspect}" unless @ftype
      @ftype == type
    end

    # Dynamic checkers
    %w(directory file symlink).each do |k|
      define_method "#{k}?" do
        file_type_is?(k.to_sym)
      end
    end

    def name_is_directory? #:nodoc:all
      @name.end_with?('/')
    end

    def local_entry_offset #:nodoc:all
      local_header_offset + @local_header_size
    end

    def name_size
      @name ? @name.bytesize : 0
    end

    def extra_size
      @extra ? @extra.local_size : 0
    end

    def comment_size
      @comment ? @comment.bytesize : 0
    end

    def calculate_local_header_size #:nodoc:all
      LOCAL_ENTRY_STATIC_HEADER_LENGTH + name_size + extra_size
    end

    # check before rewriting an entry (after file sizes are known)
    # that we didn't change the header size (and thus clobber file data or something)
    def verify_local_header_size!
      return if @local_header_size.nil?
      new_size = calculate_local_header_size
      raise Error, "local header size changed (#{@local_header_size} -> #{new_size})" if @local_header_size != new_size
    end

    def cdir_header_size #:nodoc:all
      CDIR_ENTRY_STATIC_HEADER_LENGTH + name_size +
        (@extra ? @extra.c_dir_size : 0) + comment_size
    end

    def next_header_offset #:nodoc:all
      local_entry_offset + compressed_size + data_descriptor_size
    end

    # Extracts entry to file dest_path (defaults to @name).
    def extract(dest_path = @name, &block)
      block ||= proc { ::RubyZip.on_exists_proc }

      if @name.squeeze('/') =~ /\.{2}(?:\/|\z)/
        puts "WARNING: skipped \"../\" path component(s) in #{@name}"
        return self
      end

      if directory? || file? || symlink?
        __send__("create_#{@ftype}", dest_path, &block)
      else
        raise "unknown file type #{inspect}"
      end

      self
    end

    def to_s
      @name
    end

    class << self
      def read_zip_short(io) # :nodoc:
        io.read(2).unpack('v')[0]
      end

      def read_zip_long(io) # :nodoc:
        io.read(4).unpack('V')[0]
      end

      def read_zip_64_long(io) # :nodoc:
        io.read(8).unpack('Q<')[0]
      end

      def read_c_dir_entry(io) #:nodoc:all
        path = if io.respond_to?(:path)
                 io.path
               else
                 io
               end
        entry = new(path)
        entry.read_c_dir_entry(io)
        entry
      rescue Error
        nil
      end

      def read_local_entry(io)
        entry = new(io)
        entry.read_local_entry(io)
        entry
      rescue Error
        nil
      end
    end

    public

    def unpack_local_entry(buf)
      @header_signature,
        @version,
        @fstype,
        @gp_flags,
        @compression_method,
        @last_mod_time,
        @last_mod_date,
        @crc,
        @compressed_size,
        @size,
        @name_length,
        @extra_length = buf.unpack('VCCvvvvVVVvv')
    end

    def read_local_entry(io) #:nodoc:all
      @local_header_offset = io.tell

      static_sized_fields_buf = io.read(::RubyZip::LOCAL_ENTRY_STATIC_HEADER_LENGTH) || ''

      unless static_sized_fields_buf.bytesize == ::RubyZip::LOCAL_ENTRY_STATIC_HEADER_LENGTH
        raise Error, 'Premature end of file. Not enough data for zip entry local header'
      end

      unpack_local_entry(static_sized_fields_buf)

      unless @header_signature == ::RubyZip::LOCAL_ENTRY_SIGNATURE
        raise ::RubyZip::Error, "Zip local header magic not found at location '#{local_header_offset}'"
      end
      set_time(@last_mod_date, @last_mod_time)

      @name = io.read(@name_length)
      extra = io.read(@extra_length)

      @name.gsub!('\\', '/')

      if extra && extra.bytesize != @extra_length
        raise ::RubyZip::Error, 'Truncated local zip entry header'
      else
        if @extra.is_a?(::RubyZip::ExtraField)
          @extra.merge(extra) if extra
        else
          @extra = ::RubyZip::ExtraField.new(extra)
        end
      end
      parse_zip64_extra(true)
      @local_header_size = calculate_local_header_size
    end

    def pack_local_entry
      zip64 = @extra['Zip64']
      [::RubyZip::LOCAL_ENTRY_SIGNATURE,
       @version_needed_to_extract, # version needed to extract
       @gp_flags, # @gp_flags                  ,
       @compression_method,
       @time.to_binary_dos_time, # @last_mod_time              ,
       @time.to_binary_dos_date, # @last_mod_date              ,
       @crc,
       (zip64 && zip64.compressed_size) ? 0xFFFFFFFF : @compressed_size,
       (zip64 && zip64.original_size) ? 0xFFFFFFFF : @size,
       name_size,
       @extra ? @extra.local_size : 0].pack('VvvvvvVVVvv')
    end

    def write_local_entry(io, rewrite = false) #:nodoc:all
      prep_zip64_extra(true)
      verify_local_header_size! if rewrite
      @local_header_offset = io.tell

      io << pack_local_entry

      io << @name
      io << @extra.to_local_bin if @extra
      @local_header_size = io.tell - @local_header_offset
    end

    def unpack_c_dir_entry(buf)
      @header_signature,
        @version, # version of encoding software
        @fstype, # filesystem type
        @version_needed_to_extract,
        @gp_flags,
        @compression_method,
        @last_mod_time,
        @last_mod_date,
        @crc,
        @compressed_size,
        @size,
        @name_length,
        @extra_length,
        @comment_length,
        _, # diskNumberStart
        @internal_file_attributes,
        @external_file_attributes,
        @local_header_offset,
        @name,
        @extra,
        @comment = buf.unpack('VCCvvvvvVVVvvvvvVV')
    end

    def set_ftype_from_c_dir_entry
      @ftype = case @fstype
               when ::RubyZip::FSTYPE_UNIX
                 @unix_perms = (@external_file_attributes >> 16) & 07777
                 case (@external_file_attributes >> 28)
                 when ::RubyZip::FILE_TYPE_DIR
                   :directory
                 when ::RubyZip::FILE_TYPE_FILE
                   :file
                 when ::RubyZip::FILE_TYPE_SYMLINK
                   :symlink
                 else
                   # best case guess for whether it is a file or not
                   # Otherwise this would be set to unknown and that entry would never be able to extracted
                   if name_is_directory?
                     :directory
                   else
                     :file
                   end
                 end
               else
                 if name_is_directory?
                   :directory
                 else
                   :file
                 end
               end
    end

    def check_c_dir_entry_static_header_length(buf)
      return if buf.bytesize == ::RubyZip::CDIR_ENTRY_STATIC_HEADER_LENGTH
      raise Error, 'Premature end of file. Not enough data for zip cdir entry header'
    end

    def check_c_dir_entry_signature
      return if header_signature == ::RubyZip::CENTRAL_DIRECTORY_ENTRY_SIGNATURE
      raise Error, "Zip local header magic not found at location '#{local_header_offset}'"
    end

    def check_c_dir_entry_comment_size
      return if @comment && @comment.bytesize == @comment_length
      raise ::RubyZip::Error, 'Truncated cdir zip entry header'
    end

    def read_c_dir_extra_field(io)
      if @extra.is_a?(::RubyZip::ExtraField)
        @extra.merge(io.read(@extra_length))
      else
        @extra = ::RubyZip::ExtraField.new(io.read(@extra_length))
      end
    end

    def read_c_dir_entry(io) #:nodoc:all
      static_sized_fields_buf = io.read(::RubyZip::CDIR_ENTRY_STATIC_HEADER_LENGTH)
      check_c_dir_entry_static_header_length(static_sized_fields_buf)
      unpack_c_dir_entry(static_sized_fields_buf)
      check_c_dir_entry_signature
      set_time(@last_mod_date, @last_mod_time)
      @name = io.read(@name_length)
      read_c_dir_extra_field(io)
      @comment = io.read(@comment_length)
      check_c_dir_entry_comment_size
      set_ftype_from_c_dir_entry
      parse_zip64_extra(false)
    end

    def file_stat(path) # :nodoc:
      if @follow_symlinks
        ::File.stat(path)
      else
        ::File.lstat(path)
      end
    end

    def get_extra_attributes_from_path(path) # :nodoc:
      return if RubyZip::RUNNING_ON_WINDOWS
      stat        = file_stat(path)
      @unix_uid   = stat.uid
      @unix_gid   = stat.gid
      @unix_perms = stat.mode & 07777
    end

    def set_unix_permissions_on_path(dest_path)
      # BUG: does not update timestamps into account
      # ignore setuid/setgid bits by default.  honor if @restore_ownership
      unix_perms_mask = 01777
      unix_perms_mask = 07777 if @restore_ownership
      ::FileUtils.chmod(@unix_perms & unix_perms_mask, dest_path) if @restore_permissions && @unix_perms
      ::FileUtils.chown(@unix_uid, @unix_gid, dest_path) if @restore_ownership && @unix_uid && @unix_gid && ::Process.egid == 0
      # File::utimes()
    end

    def set_extra_attributes_on_path(dest_path) # :nodoc:
      return unless file? || directory?

      case @fstype
      when ::RubyZip::FSTYPE_UNIX
        set_unix_permissions_on_path(dest_path)
      end
    end

    def pack_c_dir_entry
      zip64 = @extra['Zip64']
      [
        @header_signature,
        @version, # version of encoding software
        @fstype, # filesystem type
        @version_needed_to_extract, # @versionNeededToExtract           ,
        @gp_flags, # @gp_flags                          ,
        @compression_method,
        @time.to_binary_dos_time, # @last_mod_time                      ,
        @time.to_binary_dos_date, # @last_mod_date                      ,
        @crc,
        (zip64 && zip64.compressed_size) ? 0xFFFFFFFF : @compressed_size,
        (zip64 && zip64.original_size) ? 0xFFFFFFFF : @size,
        name_size,
        @extra ? @extra.c_dir_size : 0,
        comment_size,
        (zip64 && zip64.disk_start_number) ? 0xFFFF : 0, # disk number start
        @internal_file_attributes, # file type (binary=0, text=1)
        @external_file_attributes, # native filesystem attributes
        (zip64 && zip64.relative_header_offset) ? 0xFFFFFFFF : @local_header_offset,
        @name,
        @extra,
        @comment
      ].pack('VCCvvvvvVVVvvvvvVV')
    end

    def write_c_dir_entry(io) #:nodoc:all
      prep_zip64_extra(false)
      case @fstype
      when ::RubyZip::FSTYPE_UNIX
        ft = case @ftype
             when :file
               @unix_perms ||= 0644
               ::RubyZip::FILE_TYPE_FILE
             when :directory
               @unix_perms ||= 0755
               ::RubyZip::FILE_TYPE_DIR
             when :symlink
               @unix_perms ||= 0755
               ::RubyZip::FILE_TYPE_SYMLINK
             end

        unless ft.nil?
          @external_file_attributes = (ft << 12 | (@unix_perms & 07777)) << 16
        end
      end

      io << pack_c_dir_entry

      io << @name
      io << (@extra ? @extra.to_c_dir_bin : '')
      io << @comment
    end

    def ==(other)
      return false unless other.class == self.class
      # Compares contents of local entry and exposed fields
      keys_equal = %w(compression_method crc compressed_size size name extra filepath).all? do |k|
        other.__send__(k.to_sym) == __send__(k.to_sym)
      end
      keys_equal && time.dos_equals(other.time)
    end

    def <=>(other)
      to_s <=> other.to_s
    end

    # Returns an IO like object for the given ZipEntry.
    # Warning: may behave weird with symlinks.
    def get_input_stream(&block)
      if @ftype == :directory
        yield ::RubyZip::NullInputStream if block_given?
        ::RubyZip::NullInputStream
      elsif @filepath
        case @ftype
        when :file
          ::File.open(@filepath, 'rb', &block)
        when :symlink
          linkpath = ::File.readlink(@filepath)
          stringio = ::StringIO.new(linkpath)
          yield(stringio) if block_given?
          stringio
        else
          raise "unknown @file_type #{@ftype}"
        end
      else
        zis = ::RubyZip::InputStream.new(@zipfile, local_header_offset)
        zis.instance_variable_set(:@internal, true)
        zis.get_next_entry
        if block_given?
          begin
            yield(zis)
          ensure
            zis.close
          end
        else
          zis
        end
      end
    end

    def gather_fileinfo_from_srcpath(src_path) # :nodoc:
      stat   = file_stat(src_path)
      @ftype = case stat.ftype
               when 'file'
                 if name_is_directory?
                   raise ArgumentError,
                         "entry name '#{newEntry}' indicates directory entry, but " \
                             "'#{src_path}' is not a directory"
                 end
                 :file
               when 'directory'
                 @name += '/' unless name_is_directory?
                 :directory
               when 'link'
                 if name_is_directory?
                   raise ArgumentError,
                         "entry name '#{newEntry}' indicates directory entry, but " \
                             "'#{src_path}' is not a directory"
                 end
                 :symlink
               else
                 raise "unknown file type: #{src_path.inspect} #{stat.inspect}"
               end

      @filepath = src_path
      get_extra_attributes_from_path(@filepath)
    end

    def write_to_zip_output_stream(zip_output_stream) #:nodoc:all
      if @ftype == :directory
        zip_output_stream.put_next_entry(self, nil, nil, ::RubyZip::Entry::STORED)
      elsif @filepath
        zip_output_stream.put_next_entry(self, nil, nil, compression_method || ::RubyZip::Entry::DEFLATED)
        get_input_stream { |is| ::RubyZip::IOExtras.copy_stream(zip_output_stream, is) }
      else
        zip_output_stream.copy_raw_entry(self)
      end
    end

    def parent_as_string
      entry_name  = name.chomp('/')
      slash_index = entry_name.rindex('/')
      slash_index ? entry_name.slice(0, slash_index + 1) : nil
    end

    def get_raw_input_stream(&block)
      if @zipfile.respond_to?(:seek) && @zipfile.respond_to?(:read)
        yield @zipfile
      else
        ::File.open(@zipfile, 'rb', &block)
      end
    end

    def clean_up
      # By default, do nothing
    end

    private

    def set_time(binary_dos_date, binary_dos_time)
      @time = ::RubyZip::DOSTime.parse_binary_dos_format(binary_dos_date, binary_dos_time)
    rescue ArgumentError
      warn 'Invalid date/time in zip entry' if ::RubyZip.warn_invalid_date
    end

    def create_file(dest_path, _continue_on_exists_proc = proc { RubyZip.continue_on_exists_proc })
      if ::File.exist?(dest_path) && !yield(self, dest_path)
        raise ::RubyZip::DestinationFileExistsError,
              "Destination '#{dest_path}' already exists"
      end
      ::File.open(dest_path, 'wb') do |os|
        get_input_stream do |is|
          set_extra_attributes_on_path(dest_path)

          buf = ''
          while (buf = is.sysread(::RubyZip::Decompressor::CHUNK_SIZE, buf))
            os << buf
          end
        end
      end
    end

    def create_directory(dest_path)
      return if ::File.directory?(dest_path)
      if ::File.exist?(dest_path)
        if block_given? && yield(self, dest_path)
          ::FileUtils.rm_f dest_path
        else
          raise ::RubyZip::DestinationFileExistsError,
                "Cannot create directory '#{dest_path}'. " \
                    'A file already exists with that name'
        end
      end
      ::FileUtils.mkdir_p(dest_path)
      set_extra_attributes_on_path(dest_path)
    end

    # BUG: create_symlink() does not use &block
    def create_symlink(dest_path)
      stat = nil
      begin
        stat = ::File.lstat(dest_path)
      rescue Errno::ENOENT
      end

      io     = get_input_stream
      linkto = io.read

      if stat
        if stat.symlink?
          if ::File.readlink(dest_path) == linkto
            return
          else
            raise ::RubyZip::DestinationFileExistsError,
                  "Cannot create symlink '#{dest_path}'. " \
                      'A symlink already exists with that name'
          end
        else
          raise ::RubyZip::DestinationFileExistsError,
                "Cannot create symlink '#{dest_path}'. " \
                    'A file already exists with that name'
        end
      end

      ::File.symlink(linkto, dest_path)
    end

    # apply missing data from the zip64 extra information field, if present
    # (required when file sizes exceed 2**32, but can be used for all files)
    def parse_zip64_extra(for_local_header) #:nodoc:all
      return if @extra['Zip64'].nil?
      if for_local_header
        @size, @compressed_size = @extra['Zip64'].parse(@size, @compressed_size)
      else
        @size, @compressed_size, @local_header_offset = @extra['Zip64'].parse(@size, @compressed_size, @local_header_offset)
      end
    end

    def data_descriptor_size
      (@gp_flags & 0x0008) > 0 ? 16 : 0
    end

    # create a zip64 extra information field if we need one
    def prep_zip64_extra(for_local_header) #:nodoc:all
      return unless ::RubyZip.write_zip64_support
      need_zip64 = @size >= 0xFFFFFFFF || @compressed_size >= 0xFFFFFFFF
      need_zip64 ||= @local_header_offset >= 0xFFFFFFFF unless for_local_header
      if need_zip64
        @version_needed_to_extract = VERSION_NEEDED_TO_EXTRACT_ZIP64
        @extra.delete('Zip64Placeholder')
        zip64 = @extra.create('Zip64')
        if for_local_header
          # local header always includes size and compressed size
          zip64.original_size = @size
          zip64.compressed_size = @compressed_size
        else
          # central directory entry entries include whichever fields are necessary
          zip64.original_size = @size if @size >= 0xFFFFFFFF
          zip64.compressed_size = @compressed_size if @compressed_size >= 0xFFFFFFFF
          zip64.relative_header_offset = @local_header_offset if @local_header_offset >= 0xFFFFFFFF
        end
      else
        @extra.delete('Zip64')

        # if this is a local header entry, create a placeholder
        # so we have room to write a zip64 extra field afterward
        # (we won't know if it's needed until the file data is written)
        if for_local_header
          @extra.create('Zip64Placeholder')
        else
          @extra.delete('Zip64Placeholder')
        end
      end
    end
  end
end

# Copyright (C) 2002, 2003 Thomas Sondergaard
# rubyzip is free software; you can redistribute it and/or
# modify it under the terms of the ruby license.
